[Merge] HC-70-feat-house-register-page To dev

New-Ha opened the pull request:
# 📝 개요(이미지 첨부 선택)
### ContinuationModal
| 이어쓰기| 취소  |
| :---: | :---: |
| ![화면 기록 2024-08-02 오후 4 26 23](https://github.com/user-attachments/assets/1bba2af1-5ade-44b0-b609-25c66ce1de7d) | ![화면 기록 2024-08-02 오후 4 26 41](https://github.com/user-attachments/assets/77beb8fd-a1a4-40cc-b754-1944bcab65a2) |
- 사용자ID로 임지저장 된(temporary값이 0인) 게시글이 있는 지 확인하고, 있다면 모달을 띄웁니다.
```tsx
// useHouse.ts
const fetchTemporaryHouseId = async (
  userId: string,
) => {
  const { data, error } = await supabase
    .from('house')
    .select('id')
    .match({ user_id: userId, temporary: 0 });
  if (error) throw new Error('임시저장된 글 확인에 실패했습니다.');
  return data;
};
// HouseRegister.tsx
  useEffect(() => {
    const checkTemporaryHouse = async () => {
      if (!houseId) {
        const { id: tempHouseId } = await fetchTemporaryHouseId(userId);
        setModalState({
          ...
         // '취소'를 누르면 임시저장글을 삭제하고 모달을 닫습니다.
          onClickCancel: () => {
            deleteHousePost(tempHouseId);
            closeModal();
          },
          // '이어쓰기'를 선택하면 수정하도록 navigate합니다.
          onClickContinue: () => {
            navigate(`/house/edit/${tempHouseId}`);
            closeModal();
          },
        });
      }
    };
    checkTemporaryHouse();
  }, [userId, houseId, navigate]);
```

### validation Template1
|validation process|
|:---:|
|![화면 기록 2024-08-03 오후 5 24 24](https://github.com/user-attachments/assets/3d811ea8-dbac-47d6-80b6-ca94bac9592a)|
- '다음'버튼을 눌렀을 때 Template1 내 입력란의 유효성검사를 진행해 통과되었을 경우에만 넘어가도록 합니다.
- text filed가 아닌 이미지개수, 위치선택도 유효성검사 후 에러메시지를 띄우기 위해 state를 사용했습니다.
- 로직 상 onChange모드로 바로 감지가 안되는 필드는 trigger()로 감지하도록 하였습니다.

### Fetch user data
|fetching|
|:---:|
|![화면 기록 2024-08-03 오후 5 40 35](https://github.com/user-attachments/assets/16d5b9f2-7f73-4b5a-90e4-a294e4b8bc5b)|
- Template2에 쓰이는 사용자가 설정한 user_mate_style과 user_lifeStyle 데이터를 가져와 기본 값으로 설정해서 보여줍니다.
- submit시에 이부분은 supabase db를 업데이트하여 설정내용을 수정하도록 합니다.

```tsx
// user_lifeStyle, user_mate_style 값을 가져오는 로직
  const { userLifeStyleQuery, userMateStyleQuery } = useProfileData(
    userId as string,
  );
  const { data: userLifeStyleData, isSuccess: fetchedUserLifeStyle } =
    userLifeStyleQuery;
  const { data: userMateStyleData, isSuccess: fetchedUserMateStyle } =
    userMateStyleQuery;

  useEffect(() => {
    if (fetchedUserLifeStyle && fetchedUserMateStyle) {
      form.setValue('smoking', userLifeStyleData.smoking);
      form.setValue('pet', userLifeStyleData.pet);
      form.setValue('appeals', userLifeStyleData.appeals);
      form.setValue('mate_gender', userMateStyleData.mate_gender);
      form.setValue('mate_number', userMateStyleData.mate_number);
      form.setValue('mate_appeals', userMateStyleData.mate_appeals);
      form.setValue('prefer_mate_age', userMateStyleData.prefer_mate_age);
    }
  }, [ fetchedUserLifeStyle, fetchedUserMateStyle, form ]);
```
```tsx
// 프로필(user || mate)을 수정
export const useUserProfileUpdate = () => {
  const { mutate: updateUserProfile } = useMutation({
    mutationFn: async ({ dbName, data, userId }) => {
      const { error } = await supabase
        .from(dbName)
        .update(data)
        .eq('id', userId);
      if (error) throw new Error(`Profile update error: ${error.message}`);
    },
    ...,
  });
  return { updateUserProfile };
};
```

### Update logic
- 하우스 게시글 수정시에도 쓰이도록 update로직을 추가했습니다.
- isEditMode state를 사용해 수정시와 작성시를 구분합니다. → useParams로 houseId가 있는 경우(혹은 임시저장된 글을 불러와 temporaryId가 있는 경우) isEditMode가 true로 됩니다.
	- isEditMode가 ture인 경우, houseId(temporaryId)에 해당하는 데이터를 불러와 기본값으로 설정합니다.
	- submit시에 supabase에 insert가 아닌 update가 되도록 합니다.
	- 수정모드인 경우, 저장된 이미지를 가져와 렌더링될 수 있도록 MultiImageForm에 로직을 추가하였습니다.
	- 또한 이미지 수정시에 storage-postId에 저장되어 있던 이미지를 temporary폴더로 옮긴 후, 수정을 거쳐 업로드되도록 하였습니다.
```tsx
 // 수정 버튼을 통해 들어오거나 '이어쓰기'를 선택해 navigate된 경우, params(houseId)가 있으면 일단 isEditMode가 true로 됩니다.
  const { houseId } = useParams<{ houseId: string }>();
  const isEditMode = !!houseId;
```
```tsx
const onSaveHouse = async(formData, temporary) => {
  ...
// isEditMode가 true라면 수정하는 것이므로 update하도록 합니다.
  if (isEditMode) {
      updateHouse({
        houseData,
        houseId: (houseId as string) || (temporaryId as string),
      });
    } else {
     // isEditMode가 false라면 insert되도록 registHouse()를 호출합니다.
      registHouse(houseData);
      // 임시저장일 때는 houseListPage로, 완전 저장이면 해당 게시글로 이동시킵니다.
      if (temporary === 1) navigate(`/house/${houseId}`);
      else navigate(`/house`);
    }
}
```

```tsx
export const useHouseUpdate = () => {
  const { mutate: updateHouse, isPending: isUpdateHouse } = useMutation({
    mutationFn: async ({ houseData, houseId }) => {
      const { data, error } = await supabase
        .from('house')
        .update(houseData)
        .eq('id', houseId)
        .select();
      return data;
    },
    onMutate: async ({ houseData, houseId }) => {
      // mutationFn이 동작하기 전에 postId폴더에 있던 이미지를 temporary로 옮깁니다.
      // onMutate에서 비동기 동작을 수행하며 toast가 자동으로 넘어가지 않아서 강제로 삭제하는 로직이 추가되었습니다.
      const toastId = createToast('updateHouse', '게시글 수정 중', {
        autoClose: false,
      });
        const postStoragePath = `house/${houseData.user_id}/${houseId}`;
        const tempStoragePath = `house/${houseData.user_id}/temporary`;
        const imageList = await getStorageImage(postStoragePath);
        if (imageList) {
          const postImages = imageList.map(imgObj => imgObj.name);
          await moveImageStorage(postImages, postStoragePath, tempStoragePath);
          removeToast(toastId as string);
        }
    },
    ...,
    onSuccess: async data => {
        const { id, user_id, images, representative_img } = data[0];
       // house데이터 수정에 성공하면 insert때처럼 storage에서 image를 정리합니다.
        await saveImageStorage(user_id, images, id);
    },
  });
  return { updateHouse, isUpdateHouse };
};
```

### responsive design
|responsive|
|:---:|
|![화면 기록 2024-08-12 오후 11 48 06](https://github.com/user-attachments/assets/4e2c23a3-169e-4476-8440-168d6cb63dcc)|
- MultiImageForm 컴포넌트는 해당 페이지와 연관성이 너무 깊어져서 molecure → template 폴더로 이동하였습니다.
- HouseRegisterTemplate1, HouseRegisterTemplate2, MultiImageForm 모두 반응형으로 코드 수정하였습니다.
- mobile 사이즈일 때는 버튼이 페이지 상단에 위치하고 디자인이 바뀌도록 했습니다. 


# ✅ PR 유형
어떤 변경 사항이 있나요?

- [X] 새로운 기능 추가
- [ ] 버그 수정
- [X] CSS 등 사용자 UI 디자인 변경
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [ ] 코드 리팩토링
- [X] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [X] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

# 📌 PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [X] 커밋 메시지 컨벤션에 맞게 작성했습니다.  
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).
- [ ] 모든 팀원이 PR에 대한 리뷰를 진행했습니다.
- [ ] 이 코드 변경 사항은 기존 기능에 영향을 주지 않나요?
- [X] 팀원이 이해할 수 있도록 충분한 주석 및 코드 일관성을 준수했나요? (변수명, 함수명, 클래스명, 가독성 등)

# 💁‍♂️ PR 추가 정보
*<!-- PR과 관련된 기타 정보(E.G 개선사항, 리뷰어가 봐주었으면 하는 점)가 있다면 여기에 적어주세요. -->*
코드 리팩토링 중으로 내용이 계속해서 수정될 예정입니다.🌟

# 🙏 팀원요청 사항
<!-- PR관련하여 팀원들의 추가 작업이 필요한 경우 작성해주세요 -->

New-Ha committed:
🚀 [Feat]: Add front icon

New-Ha committed:
🚀 [Feat]: Detach the HouseRegisterTemplate and Change the Template1

- 디자인 변경에 맞춰 캐러셀 적용과 템플릿 분리
- HouseRegisterTemplate1 UI 구현
- house database에 floor 추가에 따른 houseFormType과 supabase.ts에 floor 추가

New-Ha committed:
🚀 [Feat]: Add icons for iconBadgeButtons

- 원하는 룸메이트 성별 선택에 쓰이는 아이콘 추가 → icon-male(남자성별아이콘), icon-female(여자성별아이콘), icon-gender-free(성별상관없음아이콘)

New-Ha committed:
🚀 [Feat]: implement HouseRegisterTemplate2 UI of the desired roommate part

- 하우스 등록의 원하는 룸메이트 부분의 UI 구현
- 원하는 나이 dualInputRange 생성

New-Ha committed:
📝 [Doc]: Remove unused code and annotation

- grid로 바꾸면서 주석처리했던 flex코드 삭제
- 사용하지 않는 import 삭제

New-Ha committed:
🚀 [Feat]: Passing useForm objects to the props of template components

New-Ha committed:
🚀 [Feat]: Defining display data in HouseRegisterTemplate

- 하우스 등록 페이지에서 쓰이는 badgeButton들을 구성할 displayData를 정의

New-Ha committed:
🚀 [Feat]: Add Accordion component

- 아코디언을 열고 닫을 삼각형 아이콘 추가

New-Ha committed:
Merge branch 'dev' into HC-67-UI-House-Register-Page

New-Ha committed:
🚀 [Feat]: Add pet option icons

- 하우스 등록 페이지의 내 프로필 수정에서 쓰이는 반려동물 뱃지 버튼에 쓰이는 아이콘들 추가

New-Ha committed:
🚀 [Feat]: Add user profile modification accordion UI

- 유저 프로필 수정 아코디언 UI 구현
- responsive하도록 아코디언 컴포넌트 수정
- 하우스 타입에 주석추가 및 수정

New-Ha committed:
Merge branch 'dev' into HC-70-Feat-House-Register-Page

New-Ha committed:
🚀 [Feat]: Modify Type Files by Changing House Data

- supabase 타입 파일 변경
- useForm에 기본 값 설정함에 따라 form field의 required_error가 동작하지 않아 삭제 및 수정

New-Ha committed:
🚀 [Feat]: Modify logic to use Form objects

- useState로 관리하던 이미지배열과 대표사진 로직을 form객체를 사용하도록 수정
- 이미지 갯수 유효성 검사 가능해짐
- 반응형으로 UI 변경
  - 컨테이너 그리드로 변경
  - 반응형으로 이미지 박스가 줄어들도록 사이즈 지정 삭제 및 수정
  - 버튼들 absolute, z-index 설정 등

New-Ha committed:
Merge branch 'dev' into HC-70-Feat-House-Register-Page

New-Ha committed:
🚀 [Feat]: Add logic to fetching user_mate_style and user_lifestyle

- useHouse: supabase에서 데이터를 페칭해 오는 fetchData함수 추가
- HouseRegister:
    - useForm의 defaultValues에 user_mate_style & user_lifestyle 내역 추가
    - 다음 버튼을 눌렀을 때 trigger()로 유효성검사하는 로직 추가
    - useQuery로 데이터 페칭해오는 로직 추가
    - useEffect로 페칭해 온 데이터로 useForm의 defaultValue값을 reset하도록 함
- HouseRegisterTemplate1: 위치를 선택하지 않았을 때 에러메시지를 띄우는 로직 추가
- HouseRegisterTemplate2: useState로 관리하던 상태를 form객체로 관리하도록 수정
- Accordion 펴진 상태가 default가 되도록 수정

New-Ha committed:
🚀 [Feat]: Add House Submit and User_Profile Update Logic

- user_profile을 수정할 수 있는 업데이트 mutation 추가
- submit 시 동작할 함수에 위 함수 추가와 submit 동작 수정

New-Ha committed:
🚚 [File]: Move MultiImageForm to Template from molecules

- supabase와 통신하는 로직이 합쳐진 컴포넌트로 재사용성이 불가능하다고 판단하여 molecules→Template폴더로 이동

New-Ha committed:
🚀 [Feat]: Add logic when used as a edit

- edit 시 가져온 house data를 이용해 이미지를 보여주고 지정하는 로직 추가

New-Ha committed:
🚀 [Feat]: Add ContinuationModal

- houseRegisterPage에서 임시저장된 글이 있을 때 뜰 모달 생성

New-Ha committed:
🚑 [Fix]: Modify message css in Continuation Modal

New-Ha committed:
🚀 [Feat]: Add removeToast

- 비동기 작업으로 인해 toast가 자동으로 닫히지 않을 경우 강제로 닫도록 하기 위해 추가

New-Ha committed:
🚑 [Fix]: Modify the minimum number of house images

- 논의 결과에 따라 최소 이미지 개수를 4개로 변경

New-Ha committed:
🚀 [Feat]: Add house update logic

- house update logic 추가
- storage의 이미지 처리로직 수정

New-Ha committed:
💄 [Design]: Modify padding & representative checked color

New-Ha committed:
🚀 [Feat]: Add update logic to house register

- profile 부분 update 로직 추가
- edit 모드일 때 house update 하도록 submit 수정
- fetch 부분을 useHouse로 옮겨 응집도 높임

New-Ha committed:
🚀 [Feat]: Modify houseRegister & houseEdit router

New-Ha committed:
🚀 [Feat]: Add continuationModal in HouseRegisterPage

- 임시 저장된 글이 있는 경우, 모달을 띄우고 이어쓰기를 선택하는 경우 임시저장된 글로 form 객체의 기본값을 채우는 로직 추가

New-Ha committed:
Merge branch 'dev' into HC-70-Feat-House-Register-Page

New-Ha committed:
🚀 [Feat]: Add delete post logic when click cancel in modal

- 모달에서 취소버튼을 눌렀을 때 임시저장되어 있던 해당 게시글을 삭제하도록 함

kwondns committed:
🚀 [Feat]: House 관련 Router 수정

- house 등록 및 수정 관련 라우터 변경
- house의 children이 아닌 하위 라우터로 생성
- houseList 임시 컴포넌트 생성 및 라우터 적용

kwondns committed:
🔨 [Chore]: Cherry-picks useHouse commits

- HC-102 브랜치에서 useHouse 관련 커밋들을 cherry-pick해옴

New-Ha committed:
🚑 [Fix]: Fixed bug that moved the carousel to the tab key

kwondns committed:
🚚 [File]: useHouseDetail, useHouse 확장자 수정

- 해당 파일은 컴포넌트를 반환하는 것이 아닌 단순 훅 반환
- 따라서 tsx -> ts 로 변환

New-Ha committed:
🚀 [Feat]: Add maximum condition to house_appeal

- 하우스 어필 최대 5개까지만 작성하도록 조건 추가

New-Ha committed:
🚑 [Fix]: Adjust image form trigger range

- 이미지 갯수 유효성검사를 위한 트리거에 이미지 배열만 검사하도록 수정

New-Ha committed:
🚀 [Feat]: Add queryKeys file to manage query keys

- 쿼리키를 한 곳에서 관리하도록 queryKeys 파일 추가

New-Ha committed:
🚀 [Feat]: Delete temporaryID useState

- 임시저장된 글의 id를 저장하던 상태를 삭제하고 fetch로 가져온 id를 바로 사용하도록 수정함

New-Ha committed:
🚀 [Feat]: Add HouseFormRow component

- HouseRegisterPage에서 쓰이는 소제목과 입력필드 형식의 레이아웃 컴포넌트 생성

New-Ha committed:
♻️ [Refactor]: Change code to use HouseFormRow component

- HouseFormRow의 container.grid 기본 스타일 추가
- 해당 컴포넌트를 사용해 House Register Template 수정

New-Ha committed:
🎨 [Style]: Modify buttons position in mobile

- 모바일 화면 넓이에서 디자인에 맞춰 버튼이 위로 올라오도록 위치 수정

New-Ha committed:
♻️ [Refactor]: Clean up unnecessary code & correct typos

- 불필요한 코드 수정 및 삭제
- houseData의 icon import 오타 수정

New-Ha committed:
🎨 [Style]: Modify buttons style

- 모바일 화면에서 버튼이 상단에 위치할 때 글자만 나타나도록 수정

New-Ha committed:
🚑 [Fix]: Resolve type error

- SignUpProfileFormType의 'gender'가 'mate_gender'로 바뀜에 따른 타입에러 해결

New-Ha committed:
🚑 [Fix]: Fix user mate type error

New-Ha committed:
📝 [Doc]: Modify error messages consistently & Add navigate

New-Ha committed:
♻️ [Refactor]: Clarify variable names and delete unnecessary parts

New-Ha committed:
♻️ [Refactor]: Move and divide navigate path

New-Ha committed:
💄 [Design]: Add sticky css to buttons

- 버튼 부분으 sticky로 움직이도록 수정

New-Ha committed:
♻️ [Refactor]:  House & User QueryKeys Factory Pattern 적용

- house: all, house_post
- user: all, user_lifestyle, user_mate_style

kwondns commented on src/components/templates/HouseRegisterTemplate1.tsx:
```diff
@@ -0,0 +1,448 @@
+            />
+          </HouseFormRow>
+          <HouseFormRow
+            title="위치"
```
![image](https://github.com/user-attachments/assets/118fceb8-c495-4704-90a3-6aa9b3e7ee90)

# 환경 simulator iphone 15 pro

모바일에서 다음 캐러셀의 버튼? 이 보이는 것 같습니다.

위치 선택된 badge가 초기 빈 값으로 들어갔을 때 지워지지 않는 경우가 있습니다.

New-Ha replied:
수정했습니다!

kwondns commented on src/components/templates/HouseRegisterTemplate1.tsx:
```diff
@@ -0,0 +1,448 @@
+            gridClassName="md:grid-cols-[12.8125rem_auto]"
+          >
+            <Container.FlexCol>
+              <LabelDualInputRange
```
![image](https://github.com/user-attachments/assets/d27fd452-1256-40db-9880-3fac2b938475)

# simulator iphone 15 pro safari (size 100%)
기간 선택 구간이 최소값일 때 맨 왼쪽이 살짝 잘려 보이는 현상이 있습니다.

최대 구간 선택 버튼이 안보이는 현상이 있습니다.

New-Ha replied:
width 관련 class 수정했습니다!

kwondns commented on src/components/pages/HouseRegister.tsx:
```diff
@@ -1,16 +1,358 @@
+              </Carousel>
+            </Container.FlexCol>
+          </Container.FlexCol>
+          <Container.FlexRow className="fixed bottom-[calc(100vh-12rem)] w-[93%] justify-between bg-bg pb-4 md:sticky md:bottom-0  md:border-t md:border-brown md:pt-[2.75rem]">
```
<img width="1286" alt="image" src="https://github.com/user-attachments/assets/47a778cd-cde2-4f95-a689-3b647d15c317">

이렇게 sticky 하위의 버튼들을 넘어서 표시되는 것이 의도 된 것이 맞나요?

New-Ha replied:
수정했습니다!

kwondns commented on src/components/pages/HouseRegister.tsx:
```diff
@@ -1,16 +1,358 @@
+      user_id: userId,
+    };
+
+    if (isEditMode) {
```
하우스 등록을 해봤는데 toast 성공까지 뜨는데 이동이 undefined로 됩니다.

<img width="1192" alt="스크린샷 2024-08-17 18 31 46" src="https://github.com/user-attachments/assets/7e21b5d3-ed14-4fb0-b5f5-82717d447e24">

New-Ha replied:
houseId를 받아오도록 navigator 수정했습니다!

kwondns commented on src/components/templates/HouseRegisterTemplate1.tsx:
```diff
@@ -0,0 +1,448 @@
+            gridClassName="md:grid-cols-[12.8125rem_auto]"
+          >
+            <Container.FlexCol>
+              <input
```
6개까지 생성되는 것이 맞나요?

<img width="794" alt="스크린샷 2024-08-17 18 31 12" src="https://github.com/user-attachments/assets/e33a4e7e-9d4f-443e-8a66-8a39bfbfa3e7">

New-Ha replied:
수정했습니다~!
<img width="381" alt="스크린샷 2024-08-18 오후 5 18 59" src="https://github.com/user-attachments/assets/4fdba1fc-35cc-40bf-973d-a66a95d2ea49">

New-Ha committed:
🚑 [Fix]: Resolve navigate undefined problems

New-Ha committed:
🚑 [Fix]: Resolve CSS problems

- sticky시 버튼 컨테이너 너비 수정
- 기간선택 너비 수정

New-Ha committed:
🚑 [Fix]: Resolve next carousel visible problem

- 모바일 화면 크기에서 다음 캐러셀이 보이는 문제 해결

New-Ha committed:
🚑 [Fix]: Resolve doesn't disappear location badge problem

- 초기 렌더링시 초기화되는 위치 뱃지가 사라지지 않는 문제 해결

New-Ha committed:
🚑 [Fix]: Modify house-appeal error trigger

- 기존 6개까지 뱃지가 생기고 5개까지만 가능하다는 에러메시지가 뜨는 방식에서 5개까지만 뱃지가 생기고 6개째 입력시도시에 에러메시지가 뜨도록 수정함

New-Ha committed:
♻️ [Refactor]: Unify screen class name

- 기존 'md'로 작성된 스크린 사이즈 지정 클래스명을 'tablet'으로 수정
- 'sm-me' 로 지정한 스크린사이즈 클래스명을 's-tablet'으로 수정

New-Ha committed:
♻️ [Refactor]: Refactoring MultiImageForm component

- useEffect 상태와 조건 수정
- supabase 업로드 로직 분리
- 숫자로 적혀있던 렌더링 코드를 상수로 선언한 변수명으로 수정해 가독성 높임

New-Ha committed:
🚚 [File]: Add HouseRegister directory in templates directory

- templates 폴더 안에 HouseRegister 폴더를 만들어 관련 템플릿을 밀집화함

New-Ha committed:
📝 [Doc]: Remove comments

New-Ha committed:
♻️ [Refactor]: Refactoring function and variable names

New-Ha committed:
Merge branch 'dev' into HC-70-Feat-House-Register-Page
